thoth/storages/solvers.py

```diff
@@ -29,3 +29,16 @@ class SolverResultsStore(ResultStorageBase):
     def get_solver_name_from_document_id(solver_document_id: str) -> str:
         """Retrieve solver name from solver's document id."""
         return solver_document_id.rsplit("-", maxsplit=2)[0]
+
+    def store_old_documents(self, graph, os_name=None, os_version=None, python_version=None):
+        """Store to be deleted solver documents to Ceph"""
+        solver_document_ids = graph.get_solver_run_document_ids_all(
+            os_version=os_version, os_name=os_name, python_version=python_version
+        )
+        self.connect()
+
+        for solver_document_id in solver_document_ids:
+            solver_document_id_requirements = tuple(solver_document_id.split("-")[1:4])
+            if solver_document_id_requirements == (os_name, os_version, python_version):
+                json_document = self.ceph.retrieve_document(document_id=solver_document_id)
+                self.ceph.store_document(document=json_document, document_id=solver_document_id)
```

Comment:
Would this code store `json_document` in a new location or in will it overwrite already existing document with the same content?

Reply:
The `store_document` method used here is actually based on the `boto3` `Object().put` method , were the document name to be stored is specified by the second argument of the `Object` constructor, formed by the prefix chosen in `purge-job/app.py` (for example with `solver_store = SolverResultsStore(prefix="old-solver-documents")`). From what I understand, unless using [bucket versioning](https://docs.aws.amazon.com/AmazonS3/latest/userguide/Versioning.html), if the file with name `"prefix" + filename"` is already present in the bucket, this operation will overwrite it.

Reply:
Let's say we want to run the purge job and make sure the old documents are moved to a new location. Pseudocode to do that (haven't tested, just and idea how to do that):

```python
from thoth.storages import GraphDatabase
from thoth.storages import CephStore
from thoth.storages import SolverResultsStore
from thoth.common import datetime2datetime_str

graph = GraphDatabase()
graph.connect()

solver_store = SolverResultsStore()
solver_store.connect()

# Move to solver-purge-<timestamp>.
target_prefix = f"{solver_store.ceph.prefix.rsplit("/", maxsplit=1)[0]}/solver-purge-{datetime2datetime_str()}/"
target_store = CephStore(prefix=target_prefix)
target_store.connect()

for solver_document_id in solver_document_ids:
    document = solver_store.retrieve_document(document_id=solver_document_id)
    target_store.store_document(document, document_id=solver_document_id)
    solver_store.delete_document(document_id=solver_document_id)  # needs to be implemented in ceph.py
    graph.remove_solver_entry(document_id=solver_document_id)  # needs to be implemented in postgres.py
```

This logic can be refactored and placed to `GraphDatabase` in postgres.py which would accept `solver_store` and `target_prefix` to move documents based on the query and delete the entry in the database (returns number of documents purged so we can eventually track it):

```python
class GraphDatabase:

    def purge_solver_documents(
        solver_store: SolverResultsStore,
        target_prefix: str,
        *,
        os_name: Optional[str] = None,
        Optional[os_version] = None,
        python_version: Optional[str] = None
    ) -> int:
        ...
```

Would this solution make sense?

Reply:
It makes sense, thank you!